code/game/machinery/shieldgen.dm

```diff
@@ -385,21 +423,145 @@
 			"<span class='notice'>You turn off \the [src].</span>", \
 			"<span class='italics'>You hear heavy droning fade out.</span>")
 		active = FALSE
-		update_activity()
 	else
 		user.visible_message("[user] turned \the [src] on.", \
 			"<span class='notice'>You turn on \the [src].</span>", \
 			"<span class='italics'>You hear heavy droning.</span>")
 		active = ACTIVE_SETUPFIELDS
-		update_activity()
 	add_fingerprint(user)
 
-/obj/machinery/shieldwallgen/on_emag(mob/user)
+/obj/machinery/power/shieldwallgen/proc/toggle()
+	if(!anchored)
+		return
+	if(!powernet)
+		return
+	if(active)
+		visible_message("<span class= 'notice'>The [src.name] hums as it powers down.</span>", \
+			"If this message is ever seen, something is wrong.", \
+			"<span class= 'notice'>You hear heavy droning fade out.</span>")
+		playsound(src, 'sound/machines/synth_no.ogg', 50, TRUE, frequency = 6120)
+		active = FALSE
+		log_game("[src] was deactivated by wire pulse at [AREACOORD(src)]")
+	else
+		visible_message("<span class= 'notice'>The [src.name] beeps as it powers up.</span>", \
+			"If this message is ever seen, something is wrong.", \
+			"<span class= 'notice'>You hear heavy droning.</span>")
+		active = ACTIVE_SETUPFIELDS
```

Comment:
Basically the biggest parts of them are the text to user, theres not really a lot of logic to offload to another proc for them to share. 

Theres just the checks to see if the proc is valid, then the setters after the text.

Since one proc is intended to be by user directly with multiple things calling user, I would have to pass this arg as well as account for the arg *not* being passed in the other.

Easier to just keep both separate imo than make a proc with 3 if statements